### ISSUE TITLE
[Enhancement] Navigation through Wii U Pro controller

### ISSUE THREAD

**Issue by cucholix** [dimok789/homebrew_launcher#28]
It could be nice being able to navigate HBL with the Wii U Pro Controller, so far you only can do that through Wiiremote and Gamepad.

**Comment by CreeperMario** [comment]
I've started looking into this already. (I'll also allow you to use buttons on the GamePad and Wii Remote, instead of having to use the touch screen or point).

**Comment by cucholix** [comment]
That's great, this will included in the official release or will be a mod?

**Comment by CreeperMario** [comment]
I'll pull request it, but it won't be official until dimok returns and merges the pull request (assuming he even does).

In the meantime, it will be available in my fork, which you can place on your SD card on top of the other one.

**Comment by cucholix** [comment]
I'm checking your fork, does it contain only source, not binaries?

**Comment by CreeperMario** [comment]
My repo is only source, but once my changes are actually ready, I can provide you with a binary to test-drive.

**Comment by cucholix** [comment]
Looking forward to it, thanks!

**Comment by Riyyi** [comment]
@CreeperMario 'or point' so the Wii remote already works to select homebrew? Anyway, I'm also looking forward to your pull request.

**Comment by CreeperMario** [comment]
Preliminary implementation of button-only controls
* You can press the up/down buttons on your D-Pad (on any connected controller) to cycle through your entire homebrew list (not just the ones on screen).
* You can then press A to open the information dialog for the currently selected homebrew app.
* On the information dialog, you can press A to launch the program, or B to close the dialog and return to the list of apps.
* There are currently no visual indicators as to which homebrew app is currently selected. I will fix this at a later point.

This homebrew_launcher.elf can replace the one on your SD card in sd:/wiiu/apps/homebrew_launcher
[homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/752748/homebrew_launcher.elf.zip)

**Comment by cucholix** [comment]
Thank you works great with WUPC :)

**Comment by ddurdle** [comment]
Does this fork still require a gamepad powered on?  I know the goal of the fork is to allow for other controllers to control the interface, but at it's foundation, regualr HBL won't launch without a gamepad syned and connected to the system regardless of what controller you are actually using at launch time.

**Comment by zaersc** [comment]
Hi CreeperMario, do you know if a new HBL channel title installer is needed to get your changes for pro controller support to work?  I tried just replacing the elf on my SD, but doesn't seem any different from the official 2.1 release (A button doesn't do anything, up/down just switches pages).  I launch HBL channel to get into hbl.  Thanks for your work, cool feature.

**Comment by CreeperMario** [comment]
Oh, geez. I forgot about this.

These changes currently only take effect in the ELF version. The channel edition has not been changed. I'll work on that when I have time.

**Comment by oneoofthose** [comment]
looking foward to the channel version always annoying to have to grab the gamepad

**Comment by CreeperMario** [comment]
Okay, I think I figured it out.

* You can now press the + button on the GamePad to switch into DPAD mode.
* In this DPAD mode, the GamePad's touch screen will stop responding, and a pointer will show up on screen. The user can then use the DPAD to move this pointer around on the screen, and use the A button to trigger an action.
* Pressing + again will revert to touch screen mode.
* Note that due to the way that this works, in DPAD mode, the actions of the DPAD and A buttons are overridden, so their previous actions will no longer work. However, in the case of HBL, this is not an issue.
* The same process can be performed on Wii Remote controllers, which is useful for those who want to use a Wii Remote but do not have a functioning sensor bar.
* Theoretically, users should also be able to use Classic Controllers and Wii U Pro Controllers in the same manner, however I am not in possession of these devices, and so cannot guarantee that this works.
* I added a new image file which represents the GamePad's pointer icon.

Once again, simply replace the homebrew_launcher.elf file on your SD card with this one to test these changes out.
[homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/1150853/homebrew_launcher.elf.zip)

**Comment by CreeperMario** [comment]
Here's a channel version as well. Extract this archive and put the contents inside the sd:/install folder, and install them using WUPInstaller Y-Mod or WUPInstaller GX2.

Note that if you already have the Homebrew Launcher channel installed, you should uninstall it first, before installing my version, otherwise some weird caching issues may occur, resulting in the channel being broken.

[hblout.zip](https://github.com/dimok789/homebrew_launcher/files/1150990/hblout.zip)

**Comment by cucholix** [comment]
Hotfix works perfect, thanks!

**Comment by westonlast** [comment]
The hotfix version of the channel failed to load.  The Homebrew Launcher banner displays, and then the screen freezes black.

**Comment by CreeperMario** [comment]
> The hotfix version of the channel failed to load. The Homebrew Launcher banner displays, and then the screen freezes black.

Did you delete the original Homebrew Launcher from your console (System Settings -> Data Management) before installing this one? I've found that installing new versions of titles on top of existing ones is problematic sometimes.

Try deleting the Homebrew Launcher and reinstalling it.

**Comment by westonlast** [comment]
I didn't have a previous Homebrew Launcher installed.  I use JsTypeHax which starts your custom ELF Homebrew Launcher loaded into Mii Maker.  Then I applied the signature patch.  Then I installed your channel version of Homebrew Launcher.  Black screen after banner.

**Comment by LinerFWZ** [comment]
> Okay, I think I figured it out.
> 
> * You can now press the + button on the GamePad to switch into DPAD mode.
> * In this DPAD mode, the GamePad's touch screen will stop responding, and a pointer will show up on screen. The user can then use the DPAD to move this pointer around on the screen, and use the A button to trigger an action.
> * Pressing + again will revert to touch screen mode.
> * Note that due to the way that this works, in DPAD mode, the actions of the DPAD and A buttons are overridden, so their previous actions will no longer work. However, in the case of HBL, this is not an issue.
> * The same process can be performed on Wii Remote controllers, which is useful for those who want to use a Wii Remote but do not have a functioning sensor bar.
> * Theoretically, users should also be able to use Classic Controllers and Wii U Pro Controllers in the same manner, however I am not in possession of these devices, and so cannot guarantee that this works.
> * I added a new image file which represents the GamePad's pointer icon.
> 
> Once again, simply replace the homebrew_launcher.elf file on your SD card with this one to test these changes out.
> [homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/1150853/homebrew_launcher.elf.zip)
> 
> EDIT: Hotfix update to correctly enable support for the Classic/Pro controllers.
> [homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/1151621/homebrew_launcher.elf.zip)

My wiiu has some problem，I can't click "load" or "back" button when my open hbl
Very thanks to you,your mod save my device!

**Comment by Maschell** [comment]
See https://github.com/dimok789/homebrew_launcher/pull/35#issuecomment-449742815

**Comment by Sean-Amor** [comment]
> > Okay, I think I figured it out.
> > 
> > * You can now press the + button on the GamePad to switch into DPAD mode.
> > * In this DPAD mode, the GamePad's touch screen will stop responding, and a pointer will show up on screen. The user can then use the DPAD to move this pointer around on the screen, and use the A button to trigger an action.
> > * Pressing + again will revert to touch screen mode.
> > * Note that due to the way that this works, in DPAD mode, the actions of the DPAD and A buttons are overridden, so their previous actions will no longer work. However, in the case of HBL, this is not an issue.
> > * The same process can be performed on Wii Remote controllers, which is useful for those who want to use a Wii Remote but do not have a functioning sensor bar.
> > * Theoretically, users should also be able to use Classic Controllers and Wii U Pro Controllers in the same manner, however I am not in possession of these devices, and so cannot guarantee that this works.
> > * I added a new image file which represents the GamePad's pointer icon.
> > 
> > Once again, simply replace the homebrew_launcher.elf file on your SD card with this one to test these changes out.
> > [homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/1150853/homebrew_launcher.elf.zip)
> > EDIT: Hotfix update to correctly enable support for the Classic/Pro controllers.
> > [homebrew_launcher.elf.zip](https://github.com/dimok789/homebrew_launcher/files/1151621/homebrew_launcher.elf.zip)
> 
> My wiiu has some problem，I can't use touchpad to click the "load" or "back" button when I open hbl
> Very thanks to you,your mod save my device!

Is this safe to use still? I would love to not have to use drc-sim to load up a cfw to play an installed game. (I do not have a gamepad)

**Comment by CreeperMario** [comment]
It should still work, but please keep in mind that this only changes the Homebrew Launcher, Haxchi or the web browser may still require a GamePad depending on how your system is set up.

**Comment by Sean-Amor** [comment]
Good point, I do the web browser hack to boot Homebrew with drc-sim server
and a phone as the pad. I guess I should either do the haxchi method, or
coldboot which I guess still requires haxchi initially.

**Comment by CHIPSSSSAHOYYYYY** [comment]
@CreeperMario this does not work for me btw. I am new to modding.
I am using the elf that you posted
